hydroflow_lang/src/graph/flat_graph.rs

```diff
@@ -322,197 +322,288 @@ impl FlatGraphBuilder {
+                                if is_hard {
+                                    Level::Error
+                                } else {
+                                    Level::Warning
+                                },
+                                message,
+                            ));
+                        }
+                        out_of_range
+                    }
+
+                    let inn_degree = self.flat_graph.graph.degree_in(node_id);
+                    let _ = emit_arity_error(
+                        operator,
+                        true,
+                        true,
+                        inn_degree,
+                        op_constraints.hard_range_inn,
+                        &mut self.diagnostics,
+                    ) || emit_arity_error(
+                        operator,
+                        true,
+                        false,
+                        inn_degree,
+                        op_constraints.soft_range_inn,
+                        &mut self.diagnostics,
+                    );
+
+                    let out_degree = self.flat_graph.graph.degree_out(node_id);
+                    let _ = emit_arity_error(
+                        operator,
+                        false,
+                        true,
+                        out_degree,
+                        op_constraints.hard_range_out,
+                        &mut self.diagnostics,
+                    ) || emit_arity_error(
+                        operator,
+                        false,
+                        false,
+                        out_degree,
+                        op_constraints.soft_range_out,
+                        &mut self.diagnostics,
+                    );
+
+                    fn emit_port_error<'a>(
+                        operator_span: Span,
+                        expected_ports_fn: Option<fn() -> PortListSpec>,
+                        actual_ports_iter: impl Iterator<Item = &'a PortIndexValue>,
+                        input_output: &'static str,
+                        diagnostics: &mut Vec<Diagnostic>,
+                    ) {
+                        let Some(expected_ports_fn) = expected_ports_fn else {
+                            return;
+                        };
+                        let PortListSpec::Fixed(expected_ports) = (expected_ports_fn)() else {
+                            // Separate check inside of `demux` special case.
+                            return;
+                        };
+                        let expected_ports: Vec<_> = expected_ports.into_iter().collect();
+
+                        // Reject unexpected ports.
+                        let ports: BTreeSet<_> = actual_ports_iter
+                            .inspect(|port_index_value| {
+                                if !expected_ports.iter().any(|port_index| {
+                                    port_index_value == &&port_index.clone().into()
+                                }) {
```

Comment:
Not clear on a quick read what you've discarded here

Reply:
cleaned up